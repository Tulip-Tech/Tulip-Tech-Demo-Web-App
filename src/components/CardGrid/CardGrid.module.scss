.cell {
  display: inline-block;
  padding: 4px;
}

.container {
  margin: 0 -4px;
  text-align: left;

  @for $i from 1 through 10 {
    &.cols-#{$i} {
      > .cell {
        width: calc(100% / $i);
      }
    }
  }
}
