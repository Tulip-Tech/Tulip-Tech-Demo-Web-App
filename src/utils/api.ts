export class ApiError extends Error {
  code: number;
  message: string;

  constructor(message = '', code: number) {
    super(message);
    this.name = 'ApiError';
    this.message = message;
    this.code = code;
  }
}

/**
 * Get data
 * @param response
 */
export const getDataOrThrow = async (response: Response) => {
  const data = await response.json();

  if (!response.ok) {
    const message = `Request '${response.url}' failed with ${response.status}`;
    const error = new ApiError(data?.message || message, response.status || 500);

    throw error;
  }

  return data;
};
