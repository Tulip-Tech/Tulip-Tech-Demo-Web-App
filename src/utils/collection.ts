import type { Consent, CustomerConsent } from '#types/account';
import type { Config } from '#types/Config';
import type { GenericFormValues } from '#types/form';
import type { Playlist, PlaylistItem } from '#types/playlist';
import type { PosterAspectRatio } from '#components/Card/Card';
import { cardAspectRatios } from '#components/Card/Card';

const getFiltersFromConfig = (config: Config, playlistId: string | undefined): string[] => {
  const menuItem = config.menu.find((item) => item.contentId === playlistId);
  const filters = menuItem?.filterTags?.split(',').filter(Boolean);

  return filters || [];
};

const filterPlaylist = (playlist: Playlist, filter: string) => {
  if (!filter) return playlist;

  return {
    ...playlist,
    playlist: playlist.playlist.filter(({ tags }) => (tags ? tags.split(',').includes(filter) : false)),
  };
};

const chunk = <T>(input: T[], size: number) => {
  return input?.reduce((arr: T[][], item, idx: number) => {
    return idx % size === 0 ? [...arr, [item]] : [...arr.slice(0, -1), [...arr.slice(-1)[0], item]];
  }, []);
};

const findPlaylistImageForWidth = (playlistItem: PlaylistItem, width: number): string =>
  playlistItem.images.find((img) => img.width === width)?.src || playlistItem.image;

const generatePlaylistPlaceholder = (playlistLength: number = 15): Playlist => ({
  title: '',
  playlist: new Array(playlistLength).fill({}).map(
    (_value, index) =>
      ({
        description: '',
        duration: 0,
        feedid: '',
        image: '',
        images: [],
        shelfImage: {
          image: '',
          fallbackImage: '',
        },
        backgroundImage: {
          image: '',
          fallbackImage: '',
        },
        channelLogoImage: {
          image: '',
          fallbackImage: '',
        },
        link: '',
        genre: '',
        mediaid: `placeholder_${index}`,
        pubdate: 0,
        rating: '',
        sources: [],
        tags: '',
        title: '',
        tracks: [],
      } as PlaylistItem),
  ),
});

const formatConsentValues = (publisherConsents: Consent[] | null = [], customerConsents: CustomerConsent[] | null = []) => {
  if (!publisherConsents || !customerConsents) {
    return {};
  }
  const values: Record<string, boolean> = {};
  publisherConsents?.forEach((publisherConsent) => {
    if (customerConsents?.find((customerConsent) => customerConsent.name === publisherConsent.name && customerConsent.state === 'accepted')) {
      values[publisherConsent.name] = true;
    }
  });

  return values;
};

const extractConsentValues = (consents?: Consent[]) => {
  const values: Record<string, boolean> = {};

  if (!consents) {
    return values;
  }

  consents?.forEach((consent) => {
    values[consent.name] = consent.enabledByDefault;
  });

  return values;
};

const formatConsentsFromValues = (publisherConsents: Consent[] | null, values?: GenericFormValues) => {
  const consents: CustomerConsent[] = [];

  if (!publisherConsents || !values) return consents;

  publisherConsents.forEach((consent) => {
    consents.push({
      name: consent.name,
      version: consent.version,
      state: values.consents[consent.name] ? 'accepted' : 'declined',
    });
  });

  return consents;
};

const checkConsentsFromValues = (publisherConsents: Consent[], consents: Record<string, boolean>) => {
  const customerConsents: CustomerConsent[] = [];
  const consentsErrors: string[] = [];

  if (!publisherConsents || !consents) return { customerConsents, consentsErrors };

  publisherConsents.forEach((consent) => {
    if (consent.required && !consents[consent.name]) {
      consentsErrors.push(consent.name);
    }

    customerConsents.push({
      name: consent.name,
      version: consent.version,
      state: consents[consent.name] ? 'accepted' : 'declined',
    });
  });

  return { customerConsents, consentsErrors };
};

const deepCopy = (obj: unknown) => {
  if (Array.isArray(obj) || (typeof obj === 'object' && obj !== null)) {
    return JSON.parse(JSON.stringify(obj));
  }
  return obj;
};

const parseAspectRatio = (input: unknown) => {
  if (typeof input === 'string' && (cardAspectRatios as readonly string[]).includes(input)) return input as PosterAspectRatio;
};

const parseTilesDelta = (posterAspect?: PosterAspectRatio) => {
  if (!posterAspect) {
    return 0;
  }

  const parts = posterAspect.split(':');

  return parts.length === 2 ? Math.floor(parseInt(parts[1]) / parseInt(parts[0])) : 0;
};

export {
  getFiltersFromConfig,
  filterPlaylist,
  chunk,
  findPlaylistImageForWidth,
  generatePlaylistPlaceholder,
  formatConsentValues,
  formatConsentsFromValues,
  extractConsentValues,
  checkConsentsFromValues,
  deepCopy,
  parseAspectRatio,
  parseTilesDelta,
};
