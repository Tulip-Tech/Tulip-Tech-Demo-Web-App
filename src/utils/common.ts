import { overrideIPCookieKey } from '#test/constants';

export function debounce<T extends (...args: any[]) => void>(callback: T, wait = 200) {
  let timeout: NodeJS.Timeout | null;
  return (...args: unknown[]) => {
    if (timeout) clearTimeout(timeout);
    timeout = setTimeout(() => callback(...args), wait);
  };
}

/**
 * Parse hex color and return the RGB colors
 * @param color
 * @return {{r: number, b: number, g: number}|undefined}
 */
export function hexToRgb(color: string) {
  if (color.indexOf('#') === 0) {
    color = color.slice(1);
  }

  // convert 3-digit hex to 6-digits.
  if (color.length === 3) {
    color = color[0] + color[0] + color[1] + color[1] + color[2] + color[2];
  }

  if (color.length !== 6) {
    return undefined;
  }

  return {
    r: parseInt(color.slice(0, 2), 16),
    g: parseInt(color.slice(2, 4), 16),
    b: parseInt(color.slice(4, 6), 16),
  };
}

/**
 * Get the contrast color based on the given color
 * @param {string} color Hex or RGBA color string
 * @link {https://stackoverflow.com/a/35970186/1790728}
 * @return {string}
 */
export function calculateContrastColor(color: string) {
  const rgb = hexToRgb(color);

  if (!rgb) {
    return '';
  }

  // http://stackoverflow.com/a/3943023/112731
  return rgb.r * 0.299 + rgb.g * 0.587 + rgb.b * 0.114 > 186 ? '#000000' : '#FFFFFF';
}

// Build is either Development or Production
// Mode can be dev, jwdev, demo, test, prod, etc.
export const IS_DEVELOPMENT_BUILD = import.meta.env.DEV;
// Demo mode is used to run our firebase demo instance
export const IS_DEMO_MODE = import.meta.env.MODE === 'demo';
// Test mode is used for e2e and unit tests
export const IS_TEST_MODE = import.meta.env.MODE === 'test';
// Preview mode is used for previewing Pull Requests on github
export const IS_PREVIEW_MODE = import.meta.env.MODE === 'preview';

export function logDev(message: unknown, ...optionalParams: unknown[]) {
  if (IS_DEVELOPMENT_BUILD || IS_PREVIEW_MODE) {
    if (optionalParams.length > 0) {
      console.info(message, optionalParams);
    } else {
      console.info(message);
    }
  }
}

export function getOverrideIP() {
  if (!IS_TEST_MODE && !IS_DEVELOPMENT_BUILD && !IS_PREVIEW_MODE) {
    return undefined;
  }

  return document.cookie
    .split(';')
    .find((s) => s.trim().startsWith(`${overrideIPCookieKey}=`))
    ?.split('=')[1]
    .trim();
}

export const isTruthyCustomParamValue = (value: unknown): boolean => ['true', '1', 'yes'].includes(String(value)?.toLowerCase());

export const isFalsyCustomParamValue = (value: unknown): boolean => ['false', '0', 'no'].includes(String(value)?.toLowerCase());

export function testId(value: string | undefined) {
  return IS_DEVELOPMENT_BUILD || IS_TEST_MODE || IS_PREVIEW_MODE ? value : undefined;
}
