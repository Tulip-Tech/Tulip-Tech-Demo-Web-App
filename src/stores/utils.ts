import type { State, StateCreator } from 'zustand';
import create from 'zustand';
import { devtools, subscribeWithSelector } from 'zustand/middleware';

import { IS_DEVELOPMENT_BUILD } from '#src/utils/common';

export const createStore = <T extends State>(name: string, storeFn: StateCreator<T>) => {
  const store = subscribeWithSelector(storeFn);

  // https://github.com/pmndrs/zustand/issues/852#issuecomment-1059783350
  if (IS_DEVELOPMENT_BUILD) {
    return create(
      devtools(store, {
        name,
        anonymousActionType: `${name}/Action`,
      }),
    );
  }

  return create(store);
};
