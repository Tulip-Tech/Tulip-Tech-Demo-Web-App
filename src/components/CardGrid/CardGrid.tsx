import React, { useEffect, useState } from 'react';
import classNames from 'classnames';
import InfiniteScroll from 'react-infinite-scroller';

import styles from './CardGrid.module.scss';

import useBreakpoint, { Breakpoint, Breakpoints } from '#src/hooks/useBreakpoint';
import { isLocked } from '#src/utils/entitlements';
import Card from '#components/Card/Card';
import type { AccessModel } from '#types/Config';
import type { Playlist, PlaylistItem } from '#types/playlist';
import { parseAspectRatio, parseTilesDelta } from '#src/utils/collection';
import InfiniteScrollLoader from '#components/InfiniteScrollLoader/InfiniteScrollLoader';

const INITIAL_ROW_COUNT = 50;
const LOAD_ROWS_COUNT = 50;

const defaultCols: Breakpoints = {
  [Breakpoint.xs]: 2,
  [Breakpoint.sm]: 2,
  [Breakpoint.md]: 3,
  [Breakpoint.lg]: 4,
  [Breakpoint.xl]: 5,
};

type CardGridProps = {
  playlist: Playlist;
  onCardHover?: (item: PlaylistItem) => void;
  onCardClick: (item: PlaylistItem, playlistId?: string) => void;
  watchHistory?: { [key: string]: number };
  isLoading: boolean;
  cols?: Breakpoints;
  currentCardItem?: PlaylistItem;
  currentCardLabel?: string;
  accessModel: AccessModel;
  isLoggedIn: boolean;
  hasSubscription: boolean;
};

function CardGrid({
  playlist,
  onCardClick,
  onCardHover,
  watchHistory,
  isLoading = false,
  cols = defaultCols,
  currentCardItem,
  currentCardLabel,
  accessModel,
  isLoggedIn,
  hasSubscription,
}: CardGridProps) {
  const breakpoint: Breakpoint = useBreakpoint();
  const posterAspect = parseAspectRatio(playlist.shelfImageAspectRatio);
  const visibleTiles = cols[breakpoint] + parseTilesDelta(posterAspect);
  const [rowCount, setRowCount] = useState(INITIAL_ROW_COUNT);

  useEffect(() => {
    // reset row count when the page changes
    setRowCount(INITIAL_ROW_COUNT);
  }, [playlist.feedid]);

  const renderTile = (playlistItem: PlaylistItem) => {
    const { mediaid, title, duration, seriesId, episodeNumber, seasonNumber, shelfImage } = playlistItem;

    return (
      <div className={styles.cell} key={mediaid} role="row">
        <div role="cell">
          <Card
            title={title}
            duration={duration}
            image={shelfImage}
            progress={watchHistory ? watchHistory[mediaid] : undefined}
            seriesId={seriesId}
            episodeNumber={episodeNumber}
            seasonNumber={seasonNumber}
            onClick={() => onCardClick(playlistItem, playlistItem.feedid)}
            onHover={typeof onCardHover === 'function' ? () => onCardHover(playlistItem) : undefined}
            loading={isLoading}
            isCurrent={currentCardItem && currentCardItem.mediaid === mediaid}
            currentLabel={currentCardLabel}
            isLocked={isLocked(accessModel, isLoggedIn, hasSubscription, playlistItem)}
            posterAspect={posterAspect}
          />
        </div>
      </div>
    );
  };

  return (
    <InfiniteScroll
      pageStart={0}
      loadMore={() => setRowCount((current) => current + LOAD_ROWS_COUNT)}
      hasMore={rowCount * visibleTiles < playlist.playlist.length}
      loader={<InfiniteScrollLoader key="loader" />}
    >
      <div className={classNames(styles.container, styles[`cols-${visibleTiles}`])} role="grid">
        {playlist.playlist.slice(0, rowCount * visibleTiles).map(renderTile)}
      </div>
    </InfiniteScroll>
  );
}

export default CardGrid;
