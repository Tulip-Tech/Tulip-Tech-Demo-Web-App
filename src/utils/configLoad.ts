import merge from 'lodash.merge';

import { calculateContrastColor } from '#src/utils/common';
import loadConfig, { validateConfig } from '#src/services/config.service';
import { addScript } from '#src/utils/dom';
import { PersonalShelf, useConfigStore } from '#src/stores/ConfigStore';
import type { AccessModel, Config, Styling } from '#types/Config';
import { initializeAccount } from '#src/stores/AccountController';
import { restoreWatchHistory } from '#src/stores/WatchHistoryController';
import { initializeFavorites } from '#src/stores/FavoritesController';
import { initializeAdSchedule } from '#src/stores/ConfigController';

const CONFIG_HOST = import.meta.env.APP_API_BASE_URL;

const setCssVariables = ({ backgroundColor, highlightColor, headerBackground }: Styling) => {
  const root = document.querySelector(':root') as HTMLElement;

  if (root && backgroundColor) {
    root.style.setProperty('--body-background-color', backgroundColor);
    root.style.setProperty('--background-contrast-color', calculateContrastColor(backgroundColor));
  }

  if (root && highlightColor) {
    root.style.setProperty('--highlight-color', highlightColor);
    root.style.setProperty('--highlight-contrast-color', calculateContrastColor(highlightColor));
  }
  if (root && headerBackground) {
    root.style.setProperty('--header-background', headerBackground);
    root.style.setProperty('--header-contrast-color', calculateContrastColor(headerBackground));
  }
};

const maybeInjectAnalyticsLibrary = (config: Config) => {
  if (!config.analyticsToken) return;

  return addScript('/jwpltx.js');
};

const calculateAccessModel = (config: Config): AccessModel => {
  if (config?.integrations?.cleeng?.id) {
    return config?.integrations?.cleeng?.monthlyOffer || config?.integrations?.cleeng?.yearlyOffer ? 'SVOD' : 'AUTHVOD';
  }

  if (config?.integrations?.jwp?.clientId) {
    return config?.integrations?.jwp?.assetId ? 'SVOD' : 'AUTHVOD';
  }

  return 'AVOD';
};

export async function loadAndValidateConfig(configSource: string | undefined) {
  configSource = formatSourceLocation(configSource);

  // Explicitly set default config here as a local variable,
  // otherwise if it's a module level const, the merge below causes changes to nested properties
  const defaultConfig: Config = {
    id: '',
    siteName: '',
    description: '',
    assets: {
      banner: '/images/logo.png',
    },
    content: [],
    menu: [],
    integrations: {},
    styling: {
      footerText: '',
    },
    features: {},
  };

  if (!configSource) {
    useConfigStore.setState({ config: defaultConfig });
    throw new Error('Config not defined');
  }

  let config = await loadConfig(configSource);
  config.assets = config.assets || {};

  // make sure the banner always defaults to the JWP banner when not defined in the config
  if (!config.assets.banner) {
    config.assets.banner = defaultConfig.assets.banner;
  }

  // Store the logo right away and set css variables so the error page will be branded
  useConfigStore.setState((s) => {
    s.config.assets.banner = config.assets.banner;
  });

  setCssVariables(config.styling || {});

  config = await validateConfig(config);
  config = merge({}, defaultConfig, config);

  const accessModel = calculateAccessModel(config);

  useConfigStore.setState({
    config: config,
    accessModel,
  });

  maybeInjectAnalyticsLibrary(config);

  if (config?.integrations?.cleeng?.id && config?.integrations?.jwp?.clientId) {
    throw new Error('Invalid client integration. You cannot have both Cleeng and JWP integrations enabled at the same time.');
  }

  if (config?.integrations?.cleeng?.id || config?.integrations?.jwp?.clientId) {
    await initializeAccount();
  }

  // We only request favorites and continue_watching data if there is a corresponding item in the content section
  // and a playlist in the features section.
  // We first initialize the account otherwise if we have favorites saved as externalData and in a local storage the sections may blink
  if (config.features?.continueWatchingList && config.content.some((el) => el.type === PersonalShelf.ContinueWatching)) {
    await restoreWatchHistory();
  }
  if (config.features?.favoritesList && config.content.some((el) => el.type === PersonalShelf.Favorites)) {
    await initializeFavorites();
  }

  if (config.adSchedule) {
    await initializeAdSchedule();
  }

  return config;
}

function formatSourceLocation(source?: string) {
  if (!source) {
    return undefined;
  }

  if (source.match(/^[a-z,\d]{8}$/)) {
    return `${CONFIG_HOST}/apps/configs/${source}.json`;
  }

  return source;
}
