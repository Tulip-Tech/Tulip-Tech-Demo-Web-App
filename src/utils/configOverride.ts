import type { NavigateFunction } from 'react-router/dist/lib/hooks';

import { logDev } from '#src/utils/common';
import type { Settings } from '#src/stores/SettingsStore';

// Use local storage so the override persists until cleared
const storage = window.localStorage;

const configQueryKey = 'app-config';
const configLegacyQueryKey = 'c';

const configFileStorageKey = 'config-file-override';

export function getConfigNavigateCallback(navigate: NavigateFunction) {
  return (configSource: string) => {
    navigate(
      {
        pathname: '/',
        search: new URLSearchParams([[configQueryKey, configSource]]).toString(),
      },
      { replace: true },
    );
  };
}

export function getConfigSource(searchParams: URLSearchParams, settings: Settings | undefined) {
  if (!settings) {
    return '';
  }

  // Skip all the fancy logic below if there aren't any other options besides the default anyhow
  if (!settings.UNSAFE_allowAnyConfigSource && (settings.additionalAllowedConfigSources?.length || 0) <= 0) {
    return settings.defaultConfigSource;
  }

  const configQueryParam = searchParams.get(configQueryKey) ?? searchParams.get(configLegacyQueryKey);

  if (configQueryParam !== null) {
    // If the query param exists but the value is empty, clear the storage and allow fallback to the default config
    if (!configQueryParam) {
      storage.removeItem(configFileStorageKey);
      return settings.defaultConfigSource;
    }

    // If it's valid, store it and return it
    if (isValidConfigSource(configQueryParam, settings)) {
      storage.setItem(configFileStorageKey, configQueryParam);
      return configQueryParam;
    }

    logDev(`Invalid app-config query param: ${configQueryParam}`);
  }
  // Yes this falls through from above to look up the stored value if the query string is invalid and that's OK

  const storedSource = storage.getItem(configFileStorageKey);

  // Make sure the stored value is still valid before returning it
  if (storedSource) {
    if (isValidConfigSource(storedSource, settings)) {
      return storedSource;
    }

    logDev('Invalid stored config: ' + storedSource);
    storage.removeItem(configFileStorageKey);
  }

  return settings.defaultConfigSource;
}

export function cleanupQueryParams(searchParams: URLSearchParams, settings: Settings, configSource: string | undefined) {
  let anyTouched = false;

  // Remove the old ?c= param
  if (searchParams.has(configLegacyQueryKey)) {
    searchParams.delete(configLegacyQueryKey);
    anyTouched = true;
  }

  // If there is no valid config source or the config source equals the default, remove the ?app-config= param
  if (searchParams.has(configQueryKey) && (!configSource || configSource === settings?.defaultConfigSource)) {
    searchParams.delete(configQueryKey);
    anyTouched = true;
  }

  // If the config source is not the default and the query string isn't set right, set the ?app-config= param
  if (configSource && configSource !== settings?.defaultConfigSource && searchParams.get(configQueryKey) !== configSource) {
    searchParams.set(configQueryKey, configSource);
    anyTouched = true;
  }

  return anyTouched;
}

function isValidConfigSource(source: string, settings: Settings) {
  // Dynamic values are valid as long as they are defined
  if (settings?.UNSAFE_allowAnyConfigSource) {
    return !!source;
  }

  return (
    settings?.defaultConfigSource === source || (settings?.additionalAllowedConfigSources && settings?.additionalAllowedConfigSources.indexOf(source) >= 0)
  );
}
