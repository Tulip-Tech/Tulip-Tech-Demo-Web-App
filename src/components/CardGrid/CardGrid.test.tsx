import * as React from 'react';
import { render } from '@testing-library/react';

import CardGrid from './CardGrid';

import { generatePlaylistPlaceholder } from '#src/utils/collection';

describe('<CardGrid>', () => {
  it('renders and matches snapshot', () => {
    const placeholderData = generatePlaylistPlaceholder();
    const { container } = render(
      <CardGrid
        playlist={placeholderData}
        onCardHover={vi.fn()}
        onCardClick={vi.fn()}
        isLoading={false}
        accessModel={'SVOD'}
        isLoggedIn={true}
        hasSubscription={true}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
