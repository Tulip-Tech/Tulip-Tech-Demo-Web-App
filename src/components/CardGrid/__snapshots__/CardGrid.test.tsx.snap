// Vitest Snapshot v1

exports[`<CardGrid> > renders and matches snapshot 1`] = `
<div>
  <div>
    <div
      class="_container_7175d1 _cols-5_7175d1"
      role="grid"
    >
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <div
            aria-label="play_item"
            class="_card_d75732"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <div
            aria-label="play_item"
            class="_card_d75732"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <div
            aria-label="play_item"
            class="_card_d75732"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <div
            aria-label="play_item"
            class="_card_d75732"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <div
            aria-label="play_item"
            class="_card_d75732"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <div
            aria-label="play_item"
            class="_card_d75732"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <div
            aria-label="play_item"
            class="_card_d75732"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <div
            aria-label="play_item"
            class="_card_d75732"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <div
            aria-label="play_item"
            class="_card_d75732"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <div
            aria-label="play_item"
            class="_card_d75732"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <div
            aria-label="play_item"
            class="_card_d75732"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <div
            aria-label="play_item"
            class="_card_d75732"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <div
            aria-label="play_item"
            class="_card_d75732"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <div
            aria-label="play_item"
            class="_card_d75732"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="_cell_7175d1"
        role="row"
      >
        <div
          role="cell"
        >
          <div
            aria-label="play_item"
            class="_card_d75732"
            role="button"
            tabindex="0"
          >
            <div
              class="_poster_d75732 _aspect169_d75732"
            >
              <div
                class="_meta_d75732"
              >
                <div
                  class="_tags_d75732"
                />
              </div>
            </div>
            <div
              class="_titleContainer_d75732"
            >
              <div
                class="_title_d75732"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
